### ISSUE TITLE
Reverse P/Invoke Vector3 argument passing failure on Linux

### ISSUE THREAD

**Issue by sejongoh** [dotnet/coreclr#4607]
On Linux, the following RPInvoke case fails. The second Vector3 object, v3f32_mem0 is assigned onto stack, and its z value is wrong. The testcase passes on Windows.

Native side
```
//
// RPInvoke native call for Vector3 argument
//
typedef void (__stdcall *CallBack_RPInvoke_Vector3Arg_Unix)(
    Vector3 v3f32_xmm0, 
    float   f32_xmm2,
    float   f32_xmm3,
    float   f32_xmm4,
    float   f32_xmm5,
    float   f32_xmm6,
    float   f32_xmm7,
    float   f32_mem0,
    Vector3 v3f32_mem1,
    float   f32_mem2,
    float   f32_mem3);

EXPORT(void) __stdcall nativeCall_RPInvoke_Vector3Arg_Unix(
  CallBack_RPInvoke_Vector3Arg_Unix notify)
{
    Vector3 v1, v2;
    v1.x = 1; v1.y = 2; v1.z = 3;
    v2.x = 10; v2.y = 20; v2.z = 30;
    float f0 = 0, f1 = 1, f2 = 2, f3 = 3, f4 = 4, f5 = 5, f6 = 6, f7 = 7, f8 = 8;
    notify(
        v1,  
        f0, f1, f2, f3, f4, f5,
        f6, // mapped onto stack
        v2, 
        f7, f8);
} 
```

Managed side
```
   static void callBack_RPInvoke_Vector3Arg_Unix(
        Vector3 v3f32_xmm0, 
        float   f32_xmm2,
        float   f32_xmm3,
        float   f32_xmm4,
        float   f32_xmm5,
        float   f32_xmm6,
        float   f32_xmm7,
        float   f32_mem0,
        Vector3 v3f32_mem0,
        float   f32_mem1,
        float   f32_mem2)
    {
        result = v3f32_mem0.Z == 30;
        /*
        // sum = (1, 2, 3) dot (1, 2, 3) = 14
        float sum0 = Vector3.Dot(v3f32_xmm0, v3f32_xmm0);   
        // sum = (10, 20, 30) dot (10, 20, 30) = 1400
        float sum1 = Vector3.Dot(v3f32_mem0, v3f32_mem0);       
        // sum = (1, 2, 3) dot (10, 20, 30) = 140
        float sum2 = Vector3.Dot(v3f32_xmm0, v3f32_mem0);      
        // sum = 0 + 1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 = 36
        float sum3 = f32_xmm2 + f32_xmm3 + f32_xmm4 + f32_xmm5 + f32_xmm6 + f32_xmm7
            + f32_mem0 + f32_mem1 + f32_mem2;
        
        Console.WriteLine("callBack_RPInvoke_Vector3Arg_Unix:");
        Console.WriteLine("    {0}, {1}, {2}", v3f32_xmm0.X, v3f32_xmm0.Y, v3f32_xmm0.Z);
        Console.WriteLine("    {0}, {1}, {2}", v3f32_mem0.X, v3f32_mem0.Y, v3f32_mem0.Z);
        Console.WriteLine("    Sum0,1,2,3 = {0}, {1}, {2}, {3}", sum0, sum1, sum2, sum3);
        
        result = (sum0 == 14) && (sum1 == 1400) && (sum2 == 140) && (sum3==36);
        */
    }   
```

v3f32_mem0 is assigned (10,20,30) on the native side. 

1) Native stack right before calling the callback function
$rsp = 0x7fffffffd0b0
```
(gdb) x /32f 0x7fffffffd0b0  
; 6 is f32_mem0 and 10, 20, 30 are located at [rsp + 8, 12, 16].
0x7fffffffd0b0: 6       4.59163468e-41  10      20    
0x7fffffffd0c0: 30      0       7       4.59163468e-41
```

2) Before entering IL_STUB_ReversePInvoke, it looks like the stack arguments are copied to different stack locations. $rsp is adjusted from 0x7fffffffd0b0 to 0x7fffffffcfc8, and stack arguments are found at the new stack addresses below.

```
(gdb) print $rbp
$6 = (void *) 0x7fffffffd0a0

(gdb) print $rsp
$7 = (void *) 0x7fffffffcfc8

(gdb)  x /32f 0x7fffffffcfd0
; 6. 10, 20 are correct but v3f32_mem0.z now has 1 instead of 30.
0x7fffffffcfd0: 6       4.59163468e-41  10      20  
0x7fffffffcfe0: 1       2       0       0

```

**Comment by sejongoh** [comment]
I will check in the testcase and disable this one.

I am curious how the locations of stack arguments are changed :)

**Comment by sivarv** [comment]
@sejongoh - can you try running this test case with SIMD opt turned OFF - COMPlus_FeatureSIMD=0 ?  If the test passes, then it has something to do with SIMD optimizations.

**Comment by sejongoh** [comment]
It still fails. Is any compiler-generated code executed between calling  callBack_RPInvoke_Vector3Arg_Unix and entering IL_STUB_ReversePInvoke?

**Comment by sivarv** [comment]
AFAIK, no jitted code executes between after native code called RPInvoke and before entering IL_STUB_ReversePInvoke.  Since the same PInvoke scenario works but RPInvoke scenario fails this very well could be a VM issue in RPInvoke path.

You might want to get in touch with @janvorli on this issue.

**Comment by janvorli** [comment]
@sejongoh could you please disass the native caller function and the stub? I may be able to see what's wrong from that.

**Comment by sejongoh** [comment]
native function:
```
TEXT	SEGMENT
sum$ = 64
tv73 = 72
tv131 = 80
v3f32_xmm0$ = 112
f32_xmm2$ = 120
f32_xmm3$ = 128
f32_xmm4$ = 136
f32_xmm5$ = 144
f32_xmm6$ = 152
f32_xmm7$ = 160
f32_mem0$ = 168
v3f32_mem1$ = 176
f32_mem2$ = 184
f32_mem3$ = 192
nativeCall_PInvoke_Vector3Arg_Unix PROC
; File c:\workspaces\ws0\coreclr\tests\src\jit\simd\vector3testnative.cpp
; Line 91
$LN3:
	movss	DWORD PTR [rsp+32], xmm3
	movss	DWORD PTR [rsp+24], xmm2
	movss	DWORD PTR [rsp+16], xmm1
	mov	QWORD PTR [rsp+8], rcx
	sub	rsp, 104				; 00000068H
; Line 92
	lea	rcx, OFFSET FLAT:$SG5769
	call	printf
; Line 93
	mov	rax, QWORD PTR v3f32_xmm0$[rsp]
	cvtss2sd xmm0, DWORD PTR [rax+8]
	mov	rax, QWORD PTR v3f32_xmm0$[rsp]
	cvtss2sd xmm1, DWORD PTR [rax+4]
	mov	rax, QWORD PTR v3f32_xmm0$[rsp]
	cvtss2sd xmm2, DWORD PTR [rax]
	movsd	QWORD PTR tv73[rsp], xmm2
	movaps	xmm3, xmm0
	movd	r9, xmm3
	movaps	xmm2, xmm1
	movd	r8, xmm2
	movsd	xmm0, QWORD PTR tv73[rsp]
	movaps	xmm1, xmm0
	movd	rdx, xmm1
	lea	rcx, OFFSET FLAT:$SG5770
	call	printf
; Line 94
	cvtss2sd xmm0, DWORD PTR f32_xmm7$[rsp]
	cvtss2sd xmm1, DWORD PTR f32_xmm6$[rsp]
	cvtss2sd xmm2, DWORD PTR f32_xmm5$[rsp]
	cvtss2sd xmm3, DWORD PTR f32_xmm4$[rsp]
	cvtss2sd xmm4, DWORD PTR f32_xmm3$[rsp]
	cvtss2sd xmm5, DWORD PTR f32_xmm2$[rsp]
	movsd	QWORD PTR [rsp+48], xmm0
	movsd	QWORD PTR [rsp+40], xmm1
	movsd	QWORD PTR [rsp+32], xmm2
	movd	r9, xmm3
	movaps	xmm2, xmm4
	movd	r8, xmm2
	movaps	xmm1, xmm5
	movd	rdx, xmm1
	lea	rcx, OFFSET FLAT:$SG5771
	call	printf
; Line 96
	cvtss2sd xmm0, DWORD PTR f32_mem0$[rsp]
	movaps	xmm1, xmm0
	movd	rdx, xmm1
	lea	rcx, OFFSET FLAT:$SG5772
	call	printf
; Line 97
	mov	rax, QWORD PTR v3f32_mem1$[rsp]
	cvtss2sd xmm0, DWORD PTR [rax+8]
	mov	rax, QWORD PTR v3f32_mem1$[rsp]
	cvtss2sd xmm1, DWORD PTR [rax+4]
	mov	rax, QWORD PTR v3f32_mem1$[rsp]
	cvtss2sd xmm2, DWORD PTR [rax]
	movsd	QWORD PTR tv131[rsp], xmm2
	movaps	xmm3, xmm0
	movd	r9, xmm3
	movaps	xmm2, xmm1
	movd	r8, xmm2
	movsd	xmm0, QWORD PTR tv131[rsp]
	movaps	xmm1, xmm0
	movd	rdx, xmm1
	lea	rcx, OFFSET FLAT:$SG5773
	call	printf
; Line 98
	cvtss2sd xmm0, DWORD PTR f32_mem3$[rsp]
	cvtss2sd xmm1, DWORD PTR f32_mem2$[rsp]
	movaps	xmm2, xmm0
	movd	r8, xmm2
	movd	rdx, xmm1
	lea	rcx, OFFSET FLAT:$SG5774
	call	printf
; Line 108
	mov	rax, QWORD PTR v3f32_xmm0$[rsp]
	mov	rcx, QWORD PTR v3f32_xmm0$[rsp]
	movss	xmm0, DWORD PTR [rax]
	addss	xmm0, DWORD PTR [rcx+4]
	mov	rax, QWORD PTR v3f32_xmm0$[rsp]
	addss	xmm0, DWORD PTR [rax+8]
	addss	xmm0, DWORD PTR f32_xmm2$[rsp]
	addss	xmm0, DWORD PTR f32_xmm3$[rsp]
	addss	xmm0, DWORD PTR f32_xmm4$[rsp]
	addss	xmm0, DWORD PTR f32_xmm5$[rsp]
	addss	xmm0, DWORD PTR f32_xmm6$[rsp]
	addss	xmm0, DWORD PTR f32_xmm7$[rsp]
	addss	xmm0, DWORD PTR f32_mem0$[rsp]
	mov	rax, QWORD PTR v3f32_mem1$[rsp]
	addss	xmm0, DWORD PTR [rax]
	mov	rax, QWORD PTR v3f32_mem1$[rsp]
	addss	xmm0, DWORD PTR [rax+4]
	mov	rax, QWORD PTR v3f32_mem1$[rsp]
	addss	xmm0, DWORD PTR [rax+8]
	addss	xmm0, DWORD PTR f32_mem2$[rsp]
	addss	xmm0, DWORD PTR f32_mem3$[rsp]
	movss	DWORD PTR sum$[rsp], xmm0
; Line 110
	cvtss2sd xmm0, DWORD PTR sum$[rsp]
	movaps	xmm1, xmm0
	movd	rdx, xmm1
	lea	rcx, OFFSET FLAT:$SG5775
	call	printf
; Line 112
	movss	xmm0, DWORD PTR sum$[rsp]
; Line 113
	add	rsp, 104				; 00000068H
	ret	0
```


Stub function
```
; Total bytes of code 347, prolog size 74 for method DomainBoundILStubClass:IL_STUB_ReversePInvoke(int,struct,long,struct)
; ============================================================
; Assembly listing for method DomainBoundILStubClass:IL_STUB_ReversePInvoke(struct,float,float,float,float,float,float,float,struct,float,float)
; Emitting BLENDED_CODE for X64 CPU with SSE2
; debuggable code
; rbp based frame
; fully interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T01] (  3,   3  )  struct (16) [rbp-0x10]   do-not-enreg[SB]
;  V01 arg1         [V01,T09] (  3,   3  )   float  ->  [rbp-0x14]  
;  V02 arg2         [V02,T10] (  3,   3  )   float  ->  [rbp-0x18]  
;  V03 arg3         [V03,T11] (  3,   3  )   float  ->  [rbp-0x1C]  
;  V04 arg4         [V04,T12] (  3,   3  )   float  ->  [rbp-0x20]  
;  V05 arg5         [V05,T13] (  3,   3  )   float  ->  [rbp-0x24]  
;  V06 arg6         [V06,T14] (  3,   3  )   float  ->  [rbp-0x28]  
;  V07 arg7         [V07,T31] (  1,   1  )   float  ->  [rbp+0x10]  
;  V08 arg8         [V08,T07] (  1,   1  )  struct (16) [rbp+0x18]   do-not-enreg[SB]
;  V09 arg9         [V09,T32] (  1,   1  )   float  ->  [rbp+0x28]  
;  V10 arg10        [V10,T33] (  1,   1  )   float  ->  [rbp+0x30]  
;  V11 loc0         [V11,T08] (  1,   1  )     int  ->  [rbp-0x2C]  
;  V12 loc1         [V12,T05] (  2,   2  )  struct (16) [rbp-0x40]   do-not-enreg[SB]
;  V13 loc2         [V13,T22] (  2,   2  )   float  ->  [rbp-0x44]  
;  V14 loc3         [V14,T23] (  2,   2  )   float  ->  [rbp-0x48]  
;  V15 loc4         [V15,T24] (  2,   2  )   float  ->  [rbp-0x4C]  
;  V16 loc5         [V16,T25] (  2,   2  )   float  ->  [rbp-0x50]  
;  V17 loc6         [V17,T26] (  2,   2  )   float  ->  [rbp-0x54]  
;  V18 loc7         [V18,T27] (  2,   2  )   float  ->  [rbp-0x58]  
;  V19 loc8         [V19,T28] (  2,   2  )   float  ->  [rbp-0x5C]  
;  V20 loc9         [V20,T06] (  2,   2  )  struct (16) [rbp-0x70]   do-not-enreg[SB]
;  V21 loc10        [V21,T29] (  2,   2  )   float  ->  [rbp-0x74]  
;  V22 loc11        [V22,T30] (  2,   2  )   float  ->  [rbp-0x78]  
;  V23 tmp0         [V23    ] (  3,   3  )    long  ->  [rbp-0xD8]   do-not-enreg[X] addr-exposed
;  V24 tmp1         [V24,T04] (  2,   4  )    long  ->  [rbp-0x80]  
;  V25 tmp2         [V25,T02] (  2,   4  )     ref  ->  [rbp-0xC8]   must-init
;  V26 tmp3         [V26,T00] (  4,   8  )  struct (16) [rbp-0x90]   do-not-enreg[SB]
;  V27 tmp4         [V27    ] (  3,   6  )  struct (16) [rbp-0xA0]   do-not-enreg[XSB] addr-exposed
;  V28 tmp5         [V28,T03] (  2,   4  )     ref  ->  [rbp-0xD0]   must-init
;  V29 tmp6         [V29,T15] (  2,   4  )   float  ->  [rbp-0xA4]  
;  V30 tmp7         [V30,T16] (  2,   4  )   float  ->  [rbp-0xA8]  
;  V31 tmp8         [V31,T17] (  2,   4  )   float  ->  [rbp-0xAC]  
;  V32 tmp9         [V32,T18] (  2,   4  )   float  ->  [rbp-0xB0]  
;  V33 tmp10        [V33,T19] (  2,   4  )   float  ->  [rbp-0xB4]  
;  V34 tmp11        [V34,T20] (  2,   4  )   float  ->  [rbp-0xB8]  
;  V35 tmp12        [V35,T21] (  2,   4  )   float  ->  [rbp-0xBC]  
;  V36 OutArgs      [V36    ] (  1,   1  )  lclBlk (64) [rsp+0x00]  
;
; Lcl frame size = 288

G_M60698_IG01:
       55                   push     rbp
       4881EC20010000       sub      rsp, 288
       488DAC2420010000     lea      rbp, [rsp+120H]
       4C899528FFFFFF       mov      qword ptr [rbp-D8H], r10
       33C0                 xor      rax, rax
       48898538FFFFFF       mov      qword ptr [rbp-C8H], rax
       48898530FFFFFF       mov      qword ptr [rbp-D0H], rax
       F20F1145F0           movsd    qword ptr [rbp-10H], xmm0
       F30F114DF8           movss    dword ptr [rbp-08H], xmm1
       F30F1155EC           movss    dword ptr [rbp-14H], xmm2
       F30F115DE8           movss    dword ptr [rbp-18H], xmm3
       F30F1165E4           movss    dword ptr [rbp-1CH], xmm4
       F30F116DE0           movss    dword ptr [rbp-20H], xmm5
       F30F1175DC           movss    dword ptr [rbp-24H], xmm6
       F30F117DD8           movss    dword ptr [rbp-28H], xmm7

G_M60698_IG02:
       33FF                 xor      edi, edi
       897DD4               mov      dword ptr [rbp-2CH], edi
       90                   nop      
       488B7DF0             mov      rdi, qword ptr [rbp-10H]
       48897DC0             mov      qword ptr [rbp-40H], rdi
       8B7DF8               mov      edi, dword ptr [rbp-08H]
       897DC8               mov      dword ptr [rbp-38H], edi
       90                   nop      
       90                   nop      
       F30F1045EC           movss    xmm0, dword ptr [rbp-14H]
       F30F1145BC           movss    dword ptr [rbp-44H], xmm0
       90                   nop      
       90                   nop      
       F30F1045E8           movss    xmm0, dword ptr [rbp-18H]
       F30F1145B8           movss    dword ptr [rbp-48H], xmm0
       90                   nop      
       90                   nop      
       F30F1045E4           movss    xmm0, dword ptr [rbp-1CH]
       F30F1145B4           movss    dword ptr [rbp-4CH], xmm0
       90                   nop      
       90                   nop      
       F30F1045E0           movss    xmm0, dword ptr [rbp-20H]
       F30F1145B0           movss    dword ptr [rbp-50H], xmm0
       90                   nop      
       90                   nop      
       F30F1045DC           movss    xmm0, dword ptr [rbp-24H]
       F30F1145AC           movss    dword ptr [rbp-54H], xmm0
       90                   nop      
       90                   nop      
       F30F1045D8           movss    xmm0, dword ptr [rbp-28H]
       F30F1145A8           movss    dword ptr [rbp-58H], xmm0
       90                   nop      
       90                   nop      
       F30F104510           movss    xmm0, dword ptr [rbp+10H]
       F30F1145A4           movss    dword ptr [rbp-5CH], xmm0
       90                   nop      
       90                   nop      
       488B7D18             mov      rdi, qword ptr [rbp+18H]
       48897D90             mov      qword ptr [rbp-70H], rdi
       8B7D20               mov      edi, dword ptr [rbp+20H]
       897D98               mov      dword ptr [rbp-68H], edi
       90                   nop      
       90                   nop      
       F30F104528           movss    xmm0, dword ptr [rbp+28H]
       F30F11458C           movss    dword ptr [rbp-74H], xmm0
       90                   nop      
       90                   nop      
       F30F104530           movss    xmm0, dword ptr [rbp+30H]
       F30F114588           movss    dword ptr [rbp-78H], xmm0
       90                   nop      
       488BBD28FFFFFF       mov      rdi, qword ptr [rbp-D8H]
       B810000000           mov      eax, 16
       4863C0               movsxd   rax, eax
       488B3C07             mov      rdi, qword ptr [rdi+rax]
       488B3F               mov      rdi, gword ptr [rdi]
       488B7F08             mov      rdi, gword ptr [rdi+8]
       4889BD38FFFFFF       mov      gword ptr [rbp-C8H], rdi
       488BBD28FFFFFF       mov      rdi, qword ptr [rbp-D8H]
       B810000000           mov      eax, 16
       4863C0               movsxd   rax, eax

G_M60698_IG03:
       488B3C07             mov      rdi, qword ptr [rdi+rax]
       488B3F               mov      rdi, gword ptr [rdi]
       488B7F18             mov      rdi, qword ptr [rdi+24]
       48897D80             mov      qword ptr [rbp-80H], rdi
       488BBD38FFFFFF       mov      rdi, gword ptr [rbp-C8H]
       4889BD30FFFFFF       mov      gword ptr [rbp-D0H], rdi
       488B7DC0             mov      rdi, qword ptr [rbp-40H]
       4889BD70FFFFFF       mov      qword ptr [rbp-90H], rdi
       8B7DC8               mov      edi, dword ptr [rbp-38H]
       89BD78FFFFFF         mov      dword ptr [rbp-88H], edi
       F30F1045BC           movss    xmm0, dword ptr [rbp-44H]
       F30F11855CFFFFFF     movss    dword ptr [rbp-A4H], xmm0
       F30F1045B8           movss    xmm0, dword ptr [rbp-48H]
       F30F118558FFFFFF     movss    dword ptr [rbp-A8H], xmm0
       F30F1045B4           movss    xmm0, dword ptr [rbp-4CH]
       F30F118554FFFFFF     movss    dword ptr [rbp-ACH], xmm0
       F30F1045B0           movss    xmm0, dword ptr [rbp-50H]
       F30F118550FFFFFF     movss    dword ptr [rbp-B0H], xmm0
       F30F1045AC           movss    xmm0, dword ptr [rbp-54H]
       F30F11854CFFFFFF     movss    dword ptr [rbp-B4H], xmm0
       F30F1045A8           movss    xmm0, dword ptr [rbp-58H]
       F30F118548FFFFFF     movss    dword ptr [rbp-B8H], xmm0
       F30F1045A4           movss    xmm0, dword ptr [rbp-5CH]
       F30F118544FFFFFF     movss    dword ptr [rbp-BCH], xmm0
       488B7D90             mov      rdi, qword ptr [rbp-70H]
       4889BD60FFFFFF       mov      qword ptr [rbp-A0H], rdi
       8B7D98               mov      edi, dword ptr [rbp-68H]
       89BD68FFFFFF         mov      dword ptr [rbp-98H], edi
       F30F10458C           movss    xmm0, dword ptr [rbp-74H]
       F30F11442418         movss    dword ptr [rsp+18H], xmm0
       F30F104588           movss    xmm0, dword ptr [rbp-78H]
       F30F11442420         movss    dword ptr [rsp+20H], xmm0
       488BBD30FFFFFF       mov      rdi, gword ptr [rbp-D0H]
       F20F108570FFFFFF     movsd    xmm0, qword ptr [rbp-90H]
       F30F108D78FFFFFF     movss    xmm1, dword ptr [rbp-88H]
       F30F10955CFFFFFF     movss    xmm2, dword ptr [rbp-A4H]
       F30F109D58FFFFFF     movss    xmm3, dword ptr [rbp-A8H]
       F30F10A554FFFFFF     movss    xmm4, dword ptr [rbp-ACH]
       F30F10AD50FFFFFF     movss    xmm5, dword ptr [rbcallBack_RPInvoke_Vector3Arg_Unix:
    1, 2, 3
    10, 20, 1
    Sum0,1,2,3 = 14, 501, 53, 24
RPInvoke Vector3Arg_Unix test failed
p-B0H]
       F30F10B54CFFFFFF     movss    xmm6, dword ptr [rbp-B4H]
       F30F10BD48FFFFFF     movss    xmm7, dword ptr [rbp-B8H]
       F3440F108544FFFFFF   movss    xmm8, dword ptr [rbp-BCH]
       F3440F110424         movss    dword ptr [rsp], xmm8
       F3440F6F8560FFFFFF   movdqu   xmm8, qword ptr [rbp-A0H]
       F3440F7F442408       movdqu   qword ptr [rsp+08H], xmm8
       488B4580             mov      rax, qword ptr [rbp-80H]
       FFD0                 call     rax
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      

G_M60698_IG04:
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      
       90                   nop      

G_M60698_IG05:
       488D6500             lea      rsp, [rbp]
       5D                   pop      rbp
       C3                   ret      

; Total bytes of code 590, prolog size 79 for method DomainBoundILStubClass:IL_STUB_ReversePInvoke(struct,float,float,float,float,float,float,float,struct,float,float)
```

**Comment by janvorli** [comment]
One more thing - could you please also dump the IL of the stub?

**Comment by sejongoh** [comment]
stub IL
```
****** START compiling DomainBoundILStubClass:IL_STUB_ReversePInvoke(int,struct,long,struct) (MethodHash=7fb712e5)
OPTIONS: compCodeOpt = BLENDED_CODE
OPTIONS: compDbgCode = true
OPTIONS: compDbgInfo = false
OPTIONS: compDbgEnC  = false
OPTIONS: compProcedureSplitting   = false
OPTIONS: compProcedureSplittingEH = false
OPTIONS: Stack probing is DISABLED
IL to import:
IL_0000  16                ldc.i4.0    
IL_0001  0a                stloc.0     
IL_0002  00                nop         
IL_0003  02                ldarg.0     
IL_0004  0b                stloc.1     
IL_0005  00                nop         
IL_0006  00                nop         
IL_0007  03                ldarg.1     
IL_0008  0c                stloc.2     
IL_0009  00                nop         
IL_000a  00                nop         
IL_000b  04                ldarg.2     
IL_000c  28 03 00 00 06    call         0x6000003
IL_0011  0d                stloc.3     
IL_0012  19                ldc.i4.3    
IL_0013  0a                stloc.0     
IL_0014  00                nop         
IL_0015  00                nop         
IL_0016  05                ldarg.3     
IL_0017  13 04             stloc.s      0x4
IL_0019  00                nop         
IL_001a  28 01 00 00 06    call         0x6000001
IL_001f  1f 10             ldc.i4.s     0x10
IL_0021  58                add         
IL_0022  4d                ldind.i     
IL_0023  50                ldind.ref   
IL_0024  7b 02 00 00 04    ldfld        0x4000002
IL_0029  07                ldloc.1     
IL_002a  08                ldloc.2     
IL_002b  09                ldloc.3     
IL_002c  11 04             ldloc.s      0x4
IL_002e  28 05 00 00 06    call         0x6000005
IL_0033  1f 10             ldc.i4.s     0x10
IL_0035  58                add         
IL_0036  4d                ldind.i     
IL_0037  50                ldind.ref   
IL_0038  7b 04 00 00 04    ldfld        0x4000004
IL_003d  29 ff ff ff 11    calli        0x11FFFFFF
IL_0042  00                nop         
IL_0043  00                nop         
IL_0044  00                nop         
IL_0045  00                nop         
IL_0046  00                nop         
IL_0047  20 03 00 00 40    ldc.i4       0x40000003
IL_004c  0a                stloc.0     
IL_004d  00                nop         
IL_004e  00                nop         
IL_004f  00                nop         
IL_0050  2a                ret         
```

**Comment by janvorli** [comment]
@sejongoh the IL is unfortunately from another stub (see the parameters - IL_STUB_ReversePInvoke(int,struct,long,struct))
But I have carefully analyzed the transfers between stack slots and registers and I can see where the problem is in the assembler code of the stub.
The `Z` member that's wrong is first stored to `rbp-68h`, then to `rbp-98h`, but the final transfer to `rsp+10h` is missing.
The `rbp-68h` is second half of the `loc9` and the `rbp-98h` is second half of the `tmp4`:
```
;  V20 loc9         [V20,T06] (  2,   2  )  struct (16) [rbp-0x70]   do-not-enreg[SB]
;  V27 tmp4         [V27    ] (  3,   6  )  struct (16) [rbp-0xA0]   do-not-enreg[XSB] addr-exposed
```

**Comment by sejongoh** [comment]
@janvorli  Sorry for the wrong IL. I probably forgot to comment other test cases when getting the dump. 

I confirmed the problem. This might be another bug. The dump above was obtained with SIMD optimization disabled. SIMD is turned on by default and the testcase fails. In this case, I don't see the problem. Could you take a look at the instructions below? Please let me know if I missed anything.

```
; Total bytes of code 590, prolog size 79 for method DomainBoundILStubClass:IL_STUB_ReversePInvoke(struct,float,float,float,float,float,float,float,struct,float,float)

...
;  V08 arg8         [V08,T07] (  1,   1  )  simd12  ->  [rbp+0x18]   do-not-enreg[SB]
...
...
0F104518             movups   xmm0, xmmword ptr [rbp+18H]
0F294590             movaps   xmmword ptr [rbp-70H], xmm0
...
...
0F284590             movaps   xmm0, xmmword ptr [rbp-70H]
0F298560FFFFFF       movaps   xmmword ptr [rbp-A0H], xmm0
...
...
440F288560FFFFFF     movaps   xmm8, xmmword ptr [rbp-A0H]
440F11442408         movups   xmmword ptr [rsp+08H], xmm8
488B4580             mov      rax, qword ptr [rbp-80H]
FFD0                 call     rax
```

**Comment by sejongoh** [comment]
The problem that I originally mentioned is that stack arguments already have wrong values before entering the stub function. It looks like stack arguments on the native side are copied to lower stack locations, and v3f32_mem0 and the following arguments have wrong values.

```
; $rbp in the stub code points to 0x7fffffffcfc0.
; v3f32_mem0.z, f32_mem1, f32_mem2 have wrong values.
0x7fffffffcfc0: -nan(0x7fd0a0)  4.59163468e-41  -6.09592895e+32 4.59163468e-41  
0x7fffffffcfd0: 6       4.59163468e-41  10      20           ; f32_mem0 = 6, v3f32_mem0 has wrong values
0x7fffffffcfe0: 1       2       0       0                    ; f32_mem1 and f32_mem2 have wrong values
0x7fffffffcff0: 3       0       0       0
...
...

; $rsp in the native code points to 0x7fffffffd0b0 and all stack arguments have the correct values.
0x7fffffffd0b0: 6       4.59163468e-41  10      20           ; f32_mem0 = 6, v3f32_mem0.xy = 10, 20
0x7fffffffd0c0: 30      0       7       4.59163468e-41       ; v3f32_mem0.z = 30, f32_mem1 = 7
0x7fffffffd0d0: 8       4.59163468e-41  -107374176      8    ; f32_mem2 = 8
```

**Comment by sejongoh** [comment]
@janvorli  Could you take a look at the questions above?

**Comment by janvorli** [comment]
@sejongoh I was travelling back from Redmond to Prague, I am sorry for the delay. I am just looking into that.

**Comment by sejongoh** [comment]
@janvorli Thanks!

**Comment by janvorli** [comment]
I've found what's causing the problem. It is the value of the UMThunkMarshInfo::m_cbActualArgSize, which controls how many bytes of stack arguments get copied in the UMThunkStub assembler helper. 
In our case, this value is set to 16, while the correct value is 40. So it copies just part of the call stack - the first 16 bytes, which matches what you've dumped above.
The value of m_cbActualArgSize is set at src/vm/dllimportcallback.cpp:1496:
`m_cbActualArgSize = (pStubMD != NULL) ? pStubMD->AsDynamicMethodDesc()->GetNativeStackArgSize() : pMD->SizeOfArgStack();`
In our case, it is the result of the `pStubMD->AsDynamicMethodDesc()->GetNativeStackArgSize()`.
I can see that this value is set in ILStubState::SwapStubSignatures to value returned by pStubMD->SizeOfArgStack(). And that in turn gets the value from ArgIterator::SizeOfArgStack(). So the ArgIterator is likely broken on Unix AMD64 w.r.t. stack arg size computing. 
I am reassigning this issue to myself and I'll fix the issue.

**Comment by sivarv** [comment]
@sejongoh - Based on what you said earlier in this thread above, it looks like we have another issue when the same test is run with SIMD optimizations turned OFF?  Or is it the same issue with SIMD optimizations ON?  If it is the former, we should open another separate bug to track it.  Also it would be a good idea to run the set of tests that you have added also with SIMD optimizations OFF.

**Comment by sejongoh** [comment]
@sivarv It looks another problem. An issue is opened: [https://github.com/dotnet/coreclr/issues/4720]().

**Comment by janvorli** [comment]
I've found and fixed the bug. I'll send a PR for the fix soon.

**Comment by janvorli** [comment]
Fixed by #4729
